gateway/admin.go

```diff
@@ -21,3 +32,170 @@ func (g *GatewayAdmin) Status() interface{} {
+	var numRows int
+	var err error
+	totalRowsStmt := fmt.Sprintf(`select count(*) from %s`, r.jobTableName)
+	err = runSQL(r, totalRowsStmt, &numRows)
+	return numRows, err
+}
+
+type DSStats struct {
+	SourceNums   []SourceEvents
+	NumUsers     int
+	AvgBatchSize float64
+	TableSize    int64
+	NumRows      int
+}
+
+// first_event, last_event min--maxid to event: available in dsrange
+// Average batch size ⇒ num_events we want per ds
+// writeKey, count(*)  we want source name to count per ds
+// Num Distinct users per ds
+// Avg Event size = Table_size / (avg Batch size * Total rows) is Table_size correct measure?
+// add job status group by
+/*
+EventsBySource
+================================================================================
+│───────│────────│
+│ COUNT │ SOURCE │
+│───────│────────│
+│   448 │ tr-2   │
+│───────│────────│
+================================================================================
+NumUsers :  228
+NumUsers :  1.2589285714285714
+NumUsers :  1351680
+NumUsers :  448
+*/
+func (g *GatewayRPCHandler) GetDSStats(dsName string, result *string) error {
+	var completeErr error
+	jobTableName := prefix + dsName
+	dbHandle, err := sql.Open("postgres", jobsdb.GetConnectionString())
+	if err != nil {
+		return err
+	}
+	defer dbHandle.Close() // since this also returns an error, we can explicitly close but not doing
+	runner := &SqlRunner{dbHandle: dbHandle, jobTableName: jobTableName}
+	sources, serr := runner.getUniqueSources()
+	if serr != nil {
+		misc.AppendError("getUniqueSources", &completeErr, &serr)
+	}
+	numUsers, uerr := runner.getNumUniqueUsers()
+	if serr != nil {
+		misc.AppendError("getNumUniqueUsers", &completeErr, &uerr)
+	}
+	avgBatchSize, berr := runner.getAvgBatchSize()
+	if serr != nil {
+		misc.AppendError("getAvgBatchSize", &completeErr, &berr)
+	}
+	tableSize, tserr := runner.getTableSize()
+	if serr != nil {
+		misc.AppendError("getTableSize", &completeErr, &tserr)
+	}
+	numRows, rerr := runner.getTableRowCount()
+	if serr != nil {
+		misc.AppendError("getTableRowCount", &completeErr, &rerr)
+	}
+
+	configSubscriberLock.RLock()
+	sourcesEventToCounts := make([]SourceEvents, 0)
+	for _, sourceEvent := range sources {
+		name, found := sourceIDToNameMap[sourceEvent.Source[1:len(sourceEvent.Source)-1]]
+		if found {
+			sourcesEventToCounts = append(sourcesEventToCounts, SourceEvents{sourceEvent.Count, name})
+		}
+	}
+	configSubscriberLock.RUnlock()
+	response, err := json.MarshalIndent(DSStats{sourcesEventToCounts, numUsers, avgBatchSize, tableSize, numRows}, "", " ")
+	if err != nil {
+		*result = ""
+		misc.AppendError("MarshalIndent", &completeErr, &err)
+	} else {
+		*result = string(response)
+	}
+
+	return err
+}
+
+func runSQL(runner *SqlRunner, query string, reciever interface{}) error {
+	row := runner.dbHandle.QueryRow(query)
+	err := row.Scan(reciever)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			fmt.Println("Zero rows found")
```

Comment:
Either return nil here or don't handle this at all.

Reply:
done